add cli arg option for loading parallel extension

ablanathtanalba opened the pull request:
fixes #54 by allowing loading of an arbitrary extension alongside privacy badger in the headless browser

ghostwords commented on crawler.py:
```diff
@@ -248,6 +252,9 @@ def start_driver(self):
             opts.add_argument("--load-extension=" + new_extension_path)
 
+            if self.load_extension:
+                opts.add_extension(new_extension_path)
```
What does `new_extension_path` contain? What does `self.load_extension` contain?

ablanathtanalba commented:
Mind taking a look at the latest commit @ghostwords ? It ought to copy the source for whatever parallel extension and then copy to the docker container, but it's not quite working and I'm not sure what I'm missing

ghostwords commented:
Could you write out the steps for what you expect to happen?

ablanathtanalba commented:
@ghostwords sure thing!

In my local environment, I've added a directory for any other extensions sources to live, `/parallel-extensions/`. I'm not sure how best to just commit a directory structure change (perhaps just adding a README to that dir? or I could commit the ublock crx that I've been testing with).

The proposed changes I've made to `Dockerfile` declare an environment variable for that path, then following the same logic it takes for copying the privacy badger source, I added the line `COPY parallel-extensions $EXTENSIONS` .

Hope this makes sense!

ghostwords commented on crawler.py:
```diff
@@ -286,6 +298,15 @@ def start_driver(self):
+            # install any parallel extensions to be run alongside privacy Badger
+            if self.load_extension:
+                try:
+                    self.driver.install_addon(self.load_extension)
```
Have you tried running on Firefox? `self.driver` isn't initialized yet.

ghostwords commented on crawler.py:
```diff
@@ -248,6 +252,14 @@ def start_driver(self):
+            if self.load_extension:
+                try:
+                    opts.add_extension(self.load_extension)
+                except OSError as e:
```
If `--load-extension` was provided but it points to a non-existent file, why continue? Is this a recoverable error?

ablanathtanalba replied:
I guess if the sole intention of the scan is to have an extension run in parallel to pb, then we'd want to stop it. Otherwise it's able to continue regardless of whether or not the parallel extension loads properly. I figured logging the error was enough to cover this concern, but if you think it's necessary I'm fine adding it in

ghostwords replied:
Right, I think if you provide the `--load-extension` option, you want to run a scan with that extension, or not at all.

ablanathtanalba replied:
Cool beans, the latest update will break the scan if the `--load-extension` argument doesn't work

ghostwords commented on crawler.py:
```diff
@@ -70,6 +70,9 @@
 ap.add_argument('--log-stdout', action='store_true', default=False,
                 help='If set, log to stdout as well as log.txt')
 
+ap.add_argument('--load-extension', default=None,
```
We should log the value of `--load-extension` alongside all the other significant crawl settings.

ablanathtanalba replied:
Good call :+1: added this

ghostwords commented on Dockerfile:
```diff
@@ -33,6 +33,7 @@ ENV USER=$UNAME
 ENV HOME=/home/$USER
 ENV OUTPATH=$HOME/out/
 ENV PBPATH=$HOME/privacybadger/
+ENV EXTENSIONS=$HOME/parallel-extensions/
```
Should we add a note to the README somewhere noting how to load another extension alongside Privacy Badger when running with Docker?

ghostwords replied:
Should we add any zip and CRX files inside `parallel-extensions` to `.gitignore`?

ablanathtanalba replied:
added a note in the `2. Run A Scan` section of the README :+1:

ablanathtanalba replied:
also added all zipped and CRX files in `parallel-extensions/` to the gitignore

ghostwords commented on README.md:
````diff
@@ -67,6 +67,13 @@ See the following eff.org blog post for more information: [Giving Privacy Badger
    ```
 
+   You can load another extension to run in parallel to Privacy Badger during a scan.
+   Put the source code for the extension into the `parallel-extensions/` directory,
````
It's not source code that you want to place into `parallel-extensions`, right?

ghostwords commented on crawler.py:
```diff
@@ -212,6 +212,7 @@ def get_git_info(path):
                 "  TLDs to exclude: %s\n"
                 "  Firefox ETP: %s\n"
                 "  driver capabilities:\n\n%s\n"
+                "  parallel extensions: %s\n"
```
Let's log `--load-extension` higher up in the list, so that driver capabilities, a large object, is the final entry.

ghostwords replied:
Why "extensions"? One or none, right?

ghostwords commented on .gitignore:
```diff
@@ -1 +1,3 @@
 *.sw?
+**/parallel-extensions/*.crx
```
What's the point of the leading "**"?

ablanathtanalba replied:
It's a Windows thing iirc, though since badger sett doesn't yet work in Windows, I took it out in the recent update.

ghostwords replied:
Looking up "gitignore syntax" leads to https://git-scm.com/docs/gitignore, which says the following:

>A leading "`**`" followed by a slash means match in all directories. For example, "`**/foo`" matches file or directory "`foo`" anywhere, the same as pattern "`foo`". "`**/foo/bar`" matches file or directory "`bar`" anywhere that is directly under directory "`foo`".

Elsewhere on that page:

>a leading slash is not relevant if there is already a middle slash in the pattern

ablanathtanalba replied:
Good to know! The current version of the gitignore for this PR matches for all `.crx` and all `.zip` files in the `parallel-extensions/` directory. There's no leading slash or `**`

ghostwords commented on crawler.py:
```diff
@@ -301,6 +311,10 @@ def start_driver(self):
 
+            # install a parallel extension to run alongside pb
+            if self.load_extension:
+                self.driver.install_addon(self.load_extension, temporary=True)
```
Let's load Firefox zip extension packages, to match what we do in Chrome. So, no `temporary` flag.

ghostwords commented on README.md:
````diff
@@ -67,6 +67,15 @@ See the following eff.org blog post for more information: [Giving Privacy Badger
    ```
 
+   You can load another extension to run in parallel to Privacy Badger during a scan.
+   Use the `--load-extension` flag and pass along the filepath for the unpacked extension
````
Let's load packed everywhere for consistency. CRXs in Chrome and zips in Firefox.

ghostwords commented on .gitignore:
```diff
@@ -1 +1,3 @@
 *.sw?
+parallel-extensions/*.crx
+parallel-extensions/*.zip
```
Now that I'm testing in both browsers, I realized that Firefox packaged extensions are actually XPIs ...

ablanathtanalba replied:
That's how I've been doing it on my local machine, but figured you must know something I don't and have a differently configured environment that made this work 😆updating now to make these changes to the `.gitignore` and the `README.md`

ablanathtanalba replied:
Ok, updated :+1:

ghostwords replied:
>figured you must know something I don't

Please ask next time! You want to understand and double-check the changes you make. You don't want to assume.

ghostwords commented on README.md:
```diff
@@ -67,6 +67,15 @@ See the following eff.org blog post for more information: [Giving Privacy Badger
+   You can load another extension to run in parallel to Privacy Badger during a scan.
+   Use the `--load-extension` flag and pass along the filepath for the packed extension
+   that you want to load. For Chrome, you'd pass the filepath to the `*.crx`,
+   or with Firefox, to the zipped file.
```
to the *.xpi file

ghostwords commented on crawler.py:
```diff
@@ -301,6 +311,10 @@ def start_driver(self):
 
+            # install a parallel extension to run alongside pb
+            if self.load_extension:
+                self.driver.install_addon(self.load_extension)
```
Firefox apparently requires absolute paths (while Chrome doesn't). Could we enforce an absolute path here for Firefox? (Providing either an absolute or a relative path to `--load-extension` should work in both browsers, on Docker and without.)

ablanathtanalba replied:
I'm not sure what you mean. If the user passes along the argument `parallel-extensions/some_extension.xpi` like the option expects, wouldn't that be the absolute path in both the Docker build context and when running the python script directly?

ghostwords replied:
`parallel-extensions/some_extension.xpi` is a relative path.

`/home/daly/badger-sett/parallel-extensions/some_extension.xpi` is an absolute path.

What happens now when you provide a relative path to `--load-extension` on Firefox?

ablanathtanalba replied:
I see what you mean, now. I was considering just from the Docker build's point of view. Could be that I just ran through a bunch of updates, but I'm getting the same absolute path error when trying to run a firefox build with no parallel extension option. Just to make sure -- you're able to run a regular Docker build with Firefox?

ghostwords replied:
>you're able to run a regular Docker build with Firefox?

Yes.

ablanathtanalba commented:
I rewrote the commit history here, as it was getting a little messy.

ghostwords commented on crawler.py:
```diff
@@ -328,7 +328,8 @@ def start_driver(self):
             # loads parallel extension to run alongside pb
             if self.load_extension:
-                self.driver.install_addon(self.load_extension)
+                parallel_addon_url = os.path.join('~', self.load_extension)
```
Does this assume that your current working directory is your home directory? (Does this work in Docker?) This doesn't seem like a safe assumption.

ablanathtanalba replied:
This works in Docker for me. Is there a safer alternative you'd recommend?

ablanathtanalba replied:
Re: our conversation earlier today -- I updated the most recent commit to make any path absolute that's passed to a firefox scan
